Throw exceptions when event ID doesn't match configuration

robknight opened the pull request:
Closes https://linear.app/0xparc-pcd/issue/0XP-974/zuauth-doesnt-check-event-id-when-revealed

Fixes `authenticate` so that, if the event ID is revealed and event IDs are configured as filters, the revealed event ID must match one of the filtered values.

Also updates the tests to be more careful in identifying the kind of exception thrown.

artwyman commented on packages/lib/zuauth/src/server.ts:
```diff
@@ -31,14 +33,15 @@ export async function authenticate(
+    throw new ZuAuthAuthenticationError("ZK ticket PCD is not valid");
   }
 
   if (pcd.claim.watermark.toString() !== watermark) {
```
I think the externalNullifier should be passed in by the caller to be verified here too (if it's set at all).
In general, it should be verified by anyone who makes use of the nullifierHash.

robknight replied:
The external nullifier seems redundant if the watermark is used. What extra security or utility is it providing?

artwyman commented on packages/lib/zuauth/test/zuauth.spec.ts:
```diff
@@ -192,14 +195,16 @@ describe("zuauth should work", async function () {
+    expect(thrown).to.be.true;
   });
 
   it("should not authenticate PCDs with the wrong event ID", async function () {
```
Why is this test case not new or modified?  It seems like this test case is supposed to be testing the eventID check, but was already passing when that check was missing.  That suggests to me that there might be something wrong with the test, which was causing it to throw for some other reason.  Is that why you added the new ZuAuthAuthenticationError type?  I think it may be worth doing a check on the error message to make sure you're getting an exception specifically from the case you expect, vs. accidentally triggering a public key check instead, or something similar.

robknight replied:
The original test was broken and this one isn't.

artwyman commented on packages/lib/zuauth/src/server.ts:
```diff
@@ -52,17 +55,27 @@ export async function authenticate(
+  if (
+    eventIds.size > 0 &&
+    pcd.claim.partialTicket.eventId &&
+    !eventIds.has(pcd.claim.partialTicket.eventId)
```
If eventId is not revealed, you need to check whether `validEventIds` matches the configured set of `eventIds`.  Looking at the code elsewhere, it seems like this only happens if the number of `eventIds` is <= 20, which is the limit of the circuit.  That does seem to mean that if you have >20 and eventID isn't revealed, there's no secure guarantee.

robknight replied:
> If eventId is not revealed, you need to check whether validEventIds matches the configured set of eventIds. 

Why?

artwyman replied:
In general, all the claims should be checked against expectations, so that a prover can't fool you by sending a proof of something other than the conditions you asked for.  That's because there's no cryptographic guarantee the prover can't change the args before generating the proof.  validEventIds is simply a part of that principal.

The "If eventId is not revealed" part of my statement is simply acknowledging that checking a single revealed event ID is stricter than validEventIds, so if eventId is revealed and checked here it's okay to skip checking validEventIds.  I'd probably go ahead and check everything, though, just to be safe and clear.

See my longer comment below for more thoughts on the interactions between the different ways of checking event ID.

artwyman commented on packages/lib/zuauth/src/server.ts:
```diff
@@ -52,17 +55,27 @@ export async function authenticate(
 
+  if (
+    eventIds.size > 0 &&
+    pcd.claim.partialTicket.eventId &&
```
I think you should also check that either eventID is revealed, or validEventIDs is used and checked (see below), and throw if neither is true.  Unless you think there's a case where the user intentionally doesn't want the eventID checked (in which case why would it be in the config?).

My thought process which leads to this:
One could imagine in general checking that all the fields in `partialTicket` which should be revealed are actually revealed.  You want to avoid cases where the remote site decided to give you a proof which doesn't reveal enough data.

Probably in the general ZuAuth case that's up to the 3rd-party code using the library, which will presumably notice if any field they depend on is missing.  However for fields checked by ZuAuth (eventID, productID) the app code might reasonably assume that it doesn't need to bother looking at those fields, since ZuAuth already checked them.

A case I'm a bit worried about is this scenario:
- Proof config has >20 event IDs so no use of validEventIds
- eventID is configured to be revealed
- Malicious remote app generates a proof with eventID not revealed
- This code doesn't check anything
- App code doesn't make use of event ID (assuming that ZuAuth has verified it) and never notices.

artwyman commented on packages/lib/zuauth/src/server.ts:
```diff
@@ -52,17 +55,27 @@ export async function authenticate(
+
   if (
     productIds.size > 0 &&
     pcd.claim.partialTicket.productId &&
```
As with eventID above, I think it might be necessary to throw here if the productID isn't revealed when it should be.

artwyman reviewed (changes requested):
This is good, but thinking through the security model, I don't think it's complete yet.  Basically every field in pcd.claims should be validated against the expected configuration (unless it's purely an output field like nullifierHash).  pcd.verify() proves that the claims are proven, but doesn't rule out someone sending you a PCD proving the wrong claims.

I've made some comments on individual lines where I think there's more to check.

artwyman commented on packages/lib/zuauth/src/zuauth.ts:
```diff
@@ -135,8 +147,7 @@ export function constructZkTicketProofUrl(zuAuthArgs: ZuAuthArgs): string {
       argumentType: ArgumentTypeName.StringArray,
-      value:
-        eventIds.length !== 0 && eventIds.length <= 20 ? eventIds : undefined,
+      value: undefined,
```
This is losing a feature which we advertized at Devconnect.  How sure are you that nobody is depending on it?
Does ZKTelegram use ZuAuth?  We definitely went our of our way to make the point that logging in to DMS or posting via ZuRat didn't reveal which Devconnect subevent you attended.  It would be odd to silently undo that.

It would be possible to keep using this field when you know it's sufficient (number of eventIDs is <=20, and no productIDs are used).  That being said, I don't like that it was an implicit feature before.  It would be better if it were an explicit configuration choice by the developer to use this feature, with the library throwing an exception if the dev tries to use this feature with an overlarge list.

Note that when we get to PODTickets we can have longer lists, and also triples of (key, eventID, productID), so maybe looking ahead to that is the right way to not worry too much about getting this perfect.

robknight replied:
ZuAuth was not used at Devconnect, because it didn't exist then. The underlying code in `ZkEdDSAEventTicketPCD` and `passport-interface` was used, and still permits the use of `validEventIds`.

My thinking was that `validEventIds` is not a great feature to expose - it's confusing because it only works for event IDs and not product IDs. The more features we expose, the more we need to test and to document, and the purpose of ZuAuth is specifically to reduce the API surface area when compared to using the underlying libraries directly.

That said, I can make it an optionally-configurable feature that is off by default, which seems like the best of both worlds.

robknight replied:
Hmm. After implementing this, I remember why I didn't want to do it: the `validEventIds` model is just fundamentally not helpful for authentication in a world with multiple signers. The signer is always revealed, but unless you can determine the connection between the signer and the event ID, you can't rule out the possibility of some event ID having been used by the "wrong" signer. Passing a `validEventIds` check is just never good enough for authentication purposes, which is why we can't rely on it.

artwyman replied:
Understood.  The limitation of that feature to validEventIds does assume that the signing key and product ID are either always a single value (e.g. signing key for last year's events), or don't matter for authentication (e.g. the product IDs for devconnect).  It would be possible to check for those scenarios, or leave them up to the user to configure properly.  However given comments elsewhere I'm okay with removing this so long as it's not affecting the use cases which were built to depend on it (like ZuRat for Devconnect).  With PODTickets we can bring back a much more flexible version of this.

artwyman reviewed (changes requested):
Thanks, I like the thoroughness of the checks vs. config, particular if it's truly always the same config used for the proof request (I didn't trace all the codepaths to confirm that).

I made a few tactical suggestions in comments.  My main remaining discussion point is about what the impact of these changes might be on the assumptions of existing users/apps.  Specifically:

- Is there any case left where it's allowable to not reveal eventID and productID?  I saw when they're now enforced, but I also saw them being set to true in some new places.  I'm not familiar enough with the full scope of the code to know what unchanged cases might remain where they're not set to true.  It seems like if there are existing apps which were choosing not to reveal those fields, it's a bit surprising if our library suddenly starts revealing them.

- The removal of the validEventIDs feature is a bit worrying in the same way.  If there are apps which were depending on it, their privacy model is suddenly changing.  Also that was specifically a feature we talked about a lot at Devconnect, and used in ZKTelegram for login and ZuRat, so removing it silently seems poor from a library compatibility standpoint.  Maybe it should be left in the same state as before, where it functions only within specific size limits?  I don't to overstress it given we're going to have a much more safe and flexible alternative with PODTickets and GPCs, but I also don't like the idea of changing the assumption of existing working apps, particularly on a security feature.

Both points are things I'm wiling to defer to you on after discussion, but thought were worth raising explicitly.

robknight commented on packages/lib/zuauth/src/server.ts:
```diff
@@ -31,14 +66,27 @@ export async function authenticate(
+    }
   }
 
   const publicKeys = config.map((em) => em.publicKey);
```
I don't understand the problem here. Can you give me an example?

artwyman replied:
My config says that valid tickets to look like: (pubkeyA, eventA, productA1), (pubkeyA, eventA, productA2), (pubkeyB, eventB, productB1).

I'm presented with a ticket containing (pubkeyB, eventA, productA1).  This is not actually a valid combination because the signer for B shouldn't be signing tickets from A.  However the code here accepts it.

Alternately since there's usually only one signer, I could see a ticket (pubkey, eventB, productA1).  This is invalid because it's mixing products from A with event B.  The code here accepts it anyway.

It's an unlikely issue given there's currently only one issuer (Zupass/PODBox) and it uses unique IDs.  But in theory if the signers are independent we shouldn't trust them not to step on each other.  The code here could avoid the risk by forming a set of triples of (pubkey, eventID, productID) and checking that the presented IDs match exactly one triple.  It's a bit more complicated because of the fact that some of the fields might not be revealed, so it's not a simple set membership check.

With GPC proofs, it'll be possible to prove that the triple is in a list in a ZK way directly.

robknight commented on packages/lib/zuauth/src/server.ts:
```diff
@@ -1,10 +1,46 @@
+  keyof EdDSATicketFieldsToReveal,
+  keyof ITicketData
+> = {
+  revealAttendeeEmail: "attendeeEmail",
```
This makes logical sense, but TypeScript really wants keys into objects to have the correct types. The key type for `ITicketData` is `keyof ITicketData` and not `string`, so we can't construct a valid type by doing string manipulation.

artwyman replied:
You could do it with an `as` somewhere, I think, but fair enough to stick with type safety.

robknight commented:
I've made further modifications:

From a ZuAuth perspective, it doesn't make sense to include `validEventIds`, because in cases where the event ID is hidden it is not possible to be certain that the (hidden) matching event ID really did belong to a ticket with the expected signing key for that event ID (there could be a duplicate event ID from another signer whose tickets are also in the set of valid tickets, but not for the event ID used).

I've tightened up all of the checks that ensure that the claim matches the configuration. This is important because we want to ensure that the same configuration is used to derive the prove screen URL on the client side, and for authentication on the server-side. Any mismatch may suggest a configuration error, or use of a credential prepared with a different configuration (though the `watermark` should be sufficient to prevent credential replay attacks). Therefore we now check:
- that all fields configured to be revealed are revealed
- that all fields configured _not_ to be revealed are not revealed
- that the watermark and external nullifier match those in the configuration
- that `validEventIds` is not set

I have also added some tests, and cleaned up the implementation of existing tests. This includes checking the precise exceptions thrown for each failure case.

artwyman commented on packages/lib/zuauth/src/server.ts:
```diff
@@ -1,10 +1,73 @@
+  return (
+    isEqualEdDSAPublicKey(claim.signer, config.publicKey) &&
+    claim.partialTicket.eventId === config.eventId &&
+    !!config.productId &&
```
This means a config with no productID never matches?  Seems like you might want an OR here for the case where productID is undefined?

artwyman replied:
Ping on this comment?  Am I misreading this?

artwyman commented on packages/lib/zuauth/src/server.ts:
```diff
@@ -20,9 +83,11 @@ import {
-  config: PipelineEdDSATicketZuAuthConfig[]
+  { watermark, config, fieldsToReveal, externalNullifier }: ZuAuthArgs
 ): Promise<ZKEdDSAEventTicketPCD> {
+  /**
```
nit: I think the double `*` in `/**` here marks this as being for typedoc documentation, which doesn't seem relevant inside the body of the function.  Could be just `/*` or else `//`.

robknight replied:
If I don't use the double-asterisk, then VSCode does not insert a new asterisk at the beginning of the subsequent lines. From that I inferred a distinction between using `/* ... */` as a means of "commenting out", and `/** ... */` as a means of writing multi-line comments, but thiat is just my interpretation of VSCode's logic.

artwyman commented on packages/lib/zuauth/src/server.ts:
```diff
@@ -34,37 +99,66 @@ export async function authenticate(
+      checkIsUndefined(pcd.claim.partialTicket[fieldName], fieldName);
+    }
+  }
+
```
If you're now desupporting validEventIds, should you check here that it's required to be undefined?  I don't think it changes the security (given you're checking that all the necessary fields are being revealed), but it would catch a configuration error.

robknight replied:
I've added a check for this too.

artwyman reviewed (changes requested):
Looking good.  A few questions remaining.
I didn't review the utests in detail.

artwyman commented on packages/lib/zuauth/src/server.ts:
```diff
@@ -47,8 +47,8 @@ function claimMatchesConfiguration(
     claim.partialTicket.eventId === config.eventId &&
-    !!config.productId &&
-    claim.partialTicket.productId === config.productId
+    (config.productId === undefined ||
```
What's the semantics in PODBox if productID isn't defined?  The underlying PCD always has a productID, so is it meant to be ignored?  Is it set to be equal to the eventID?
This code implements the former (productID will be ignored) which seems fine.  We could be extra-strict and require that the productID not be revealed when the config doesn't specify it, but that seems optional
